docs/current_docs/api/clients-graphql.mdx

```diff
@@ -0,0 +1,37 @@
+---
+slug: /api/clients/graphql
+---
+import Tabs from '@theme/Tabs';
+import TabItem from '@theme/TabItem';
+
+# GraphQL Clients
+
+It's possible to call the Dagger API from [any language that supports GraphQL](https://graphql.org/code/). GraphQL has a large and growing list of client implementations in over 20 languages.
+
+This examples demonstrates how to connect to the Dagger API and run a simple pipeline in the following languages:
+
+- Rust, using the [gql_client library](https://github.com/arthurkhlghatyan/gql-client-rs) (MIT License)
+- PHP, using the [php-graphql-client library](https://github.com/mghoneimy/php-graphql-client) (MIT License)
+
+Once the client library is installed, create an API client as described below.
+
+<Tabs>
+<TabItem value="Rust">
+
+Add the following code to `src/main.rs`:
+
+```rust file=./snippets/clients-graphql/main.rs
```

Comment:
Fails with the following errors when compiled with Cargo 1.80 and rustc 1.80

```
     Locking 1 package to latest compatible version
      Adding test v0.1.0 (/home/admin/rust)
   Compiling test v0.1.0 (/home/admin/rust)
error[E0432]: unresolved import `base64`
 --> src/main.rs:1:5
  |
1 | use base64::encode;
  |     ^^^^^^ use of undeclared crate or module `base64`

error[E0432]: unresolved import `serde_json`
 --> src/main.rs:3:5
  |
3 | use serde_json::Value;
  |     ^^^^^^^^^^ use of undeclared crate or module `serde_json`

error[E0433]: failed to resolve: use of undeclared crate or module `tokio`
 --> src/main.rs:7:3
  |
7 | #[tokio::main]
  |   ^^^^^ use of undeclared crate or module `tokio`

error[E0752]: `main` function is not allowed to be `async`
 --> src/main.rs:8:1
  |
8 | async fn main() -> Result<(), Box<dyn std::error::Error>> {
  | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `main` function is not allowed to be `async`
```

@marcosnils can you please help update this code?

Reply:
here's the updated script:

```rust
use base64::{engine::general_purpose::URL_SAFE, Engine as _};
use gql_client::Client;
use serde_json::Value;
use std::collections::HashMap;
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let port = env::var("DAGGER_SESSION_PORT").expect("$DAGGER_SESSION_PORT doesn't exist");
    let token = env::var("DAGGER_SESSION_TOKEN").expect("$DAGGER_SESSION_TOKEN doesn't exist");
    let query = r#"
    query {
      container {
        from (address: "alpine:latest") {
          withExec(args:["uname", "-nrio"]) {
            stdout
          }
        }
      }
    }
 "#;

    let mut headers = HashMap::new();
    headers.insert(
        "authorization",
        format!("Basic {}", URL_SAFE.encode(format!("{}:", token))),
    );
    let client = Client::new_with_headers(format!("http://127.0.0.1:{}/query", port), headers);
    let data = client.query_unwrap::<Value>(query).await.unwrap();

    println!(
        "{}",
        data["container"]["from"]["withExec"]["stdout"]
            .as_str()
            .unwrap()
    );

    Ok(())
}
```

Here's my Cargo.toml:


```
[package]
name = "gql-client-test"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
base64 = "0.22.1"
gql_client = "1.0.7"
serde = { version = "1.0.148", features = ["derive"] }
serde_json = "1.0.125"
tokio = {version="1.22.0", features =["full"]}
```

Reply:
Fixed now